Prepare for keras-mxnet 2.2 release

sandeep-krishnamurthy opened the pull request:
# Changes
* Pull and rebase keras-team/keras v2.2.0 release
* Fix all changes on the backend, tests, examples

# Testing Done:
* All UTs / ITs are passing
* All examples are running
* Testing on GPU machines - Default usage of 1st GPU and multi_gpu_model for Sequential and Model
* Multi GPU models - Sequential / Model
 
@roywei @kalyc

sandeep-krishnamurthy commented:
updated the issue with all the changes - https://github.com/awslabs/keras-apache-mxnet/issues/92

kalyc commented on benchmark/scripts/run_mxnet_backend.sh:
```diff
@@ -32,4 +32,8 @@ echo -e "Running tests with the following config:\n$(cat ~/.keras/keras.json)"
 python $dir/run_benchmark.py  --pwd=$dir --mode="$1" --model_name="$2" --dry_run=True --inference="$3" --epochs="$4"
+=======
+python $dir/run_benchmark.py  --pwd=$dir --mode="$1" --model_name="$2" --dry_run=True --inference="$3" --epochs="$4"
+>>>>>>> Rebase to latest Keras April 20 2018 (#71)
```
please remove

sandeep-krishnamurthy replied:
oops. Thanks for pointing it out. something I missed out when rebasing. We review all other places again.

kalyc commented on benchmark/scripts/run_mxnet_backend.sh:
```diff
@@ -32,4 +32,8 @@ echo -e "Running tests with the following config:\n$(cat ~/.keras/keras.json)"
 
 dir=`pwd`
 
+<<<<<<< HEAD
```
please remove

sandeep-krishnamurthy replied:
done

kalyc commented on benchmark/scripts/run_tf_backend.sh:
```diff
@@ -32,4 +32,8 @@ echo -e "Running tests with the following config:\n$(cat ~/.keras/keras.json)"
 
 dir=`pwd`
 
+<<<<<<< HEAD
```
please remove

sandeep-krishnamurthy replied:
done

kalyc commented on benchmark/scripts/run_tf_backend.sh:
```diff
@@ -32,4 +32,8 @@ echo -e "Running tests with the following config:\n$(cat ~/.keras/keras.json)"
 python $dir/run_benchmark.py  --pwd=$dir --mode="$1" --model_name="$2" --dry_run=True --inference="$3" --epochs="$4"
+=======
+python $dir/run_benchmark.py  --pwd=$dir --mode="$1" --model_name="$2" --dry_run=True --inference="$3" --epochs="$4"
+>>>>>>> Rebase to latest Keras April 20 2018 (#71)
```
please remove

sandeep-krishnamurthy replied:
done

kalyc commented on docs/autogen.py:
```diff
@@ -418,17 +425,24 @@ def post_process_signature(signature):
 
 
+def clean_module_name(name):
+    if name.startswith('keras_applications'):
```
have you considered using regex here to replace the '_' to '.'

sandeep-krishnamurthy replied:
comes with keras-team/keras. Avoided to change.

kalyc commented on docs/templates/getting-started/faq.md:
````diff
@@ -604,6 +603,11 @@ K.set_session(sess)
 # Rest of code follows ...
 ```
 
+<<<<<<< HEAD
````
please remove

sandeep-krishnamurthy replied:
done.

kalyc commented on docs/templates/getting-started/faq.md:
```diff
@@ -604,6 +603,11 @@ K.set_session(sess)
+---
+
+=======
+>>>>>>> Rebase to latest Keras April 20 2018 (#71)
```
please remove

sandeep-krishnamurthy replied:
done

kalyc commented on docs/templates/models/about-keras-models.md:
````diff
@@ -1,35 +1,102 @@
 yaml_string = model.to_yaml()
 model = model_from_yaml(yaml_string)
 ```
+<<<<<<< HEAD
````
remove

sandeep-krishnamurthy replied:
done

kalyc commented on docs/templates/models/about-keras-models.md:
```diff
@@ -1,35 +1,102 @@
-Note: Please also see [How can I install HDF5 or h5py to save my models in Keras?](/getting-started/faq/#how-can-i-install-HDF5-or-h5py-to-save-my-models-in-Keras) in the FAQ for instructions on how to install `h5py`.
\ No newline at end of file
+Note: Please also see [How can I install HDF5 or h5py to save my models in Keras?](/getting-started/faq/#how-can-i-install-HDF5-or-h5py-to-save-my-models-in-Keras) in the FAQ for instructions on how to install `h5py`.
+>>>>>>> Rebase to latest Keras April 20 2018 (#71)
```
remove

sandeep-krishnamurthy replied:
done

kalyc commented on keras/engine/sequential.py:
```diff
@@ -0,0 +1,301 @@
+try:
+    import h5py
+except ImportError:
+    h5py = None
```
will this affect the sequential model if we dont have h5py?

sandeep-krishnamurthy replied:
h5py is installed as part of keras-mxnet pypi

kalyc replied:
Then why do we have this try catch?

sandeep-krishnamurthy replied:
part of keras. did not want to update as it is an extra check and avoid extra conflict resolution.

kalyc commented on tests/keras/engine/test_training.py:
```diff
@@ -30,54 +28,56 @@ def __len__(self):
 
 @keras_test
-def test_slice_arrays():
+def testslice_arrays():
```
please rename to test_slice_arrays()

sandeep-krishnamurthy replied:
less freedom - keras-team/keras change. Since minor change avoiding merge conflicts

kalyc commented on tests/keras/layers/convolutional_test.py:
```diff
@@ -65,15 +64,18 @@ def test_causal_dilated_conv():
                )
 
 
+@pytest.mark.skipif(K.backend() == 'mxnet',
```
why are we skipping this test? Please remove

sandeep-krishnamurthy replied:
These tests are failing with latest test parts. I wanted to request you and Lai to take a look.

kalyc replied:
I tested this change on dev branch and the conv1d test appear to pass without having to skip it for mxnet. 
There appears to be no issue with the image data format either for causal mode with conv1d - tested for both channels_first and channels_last.

sandeep-krishnamurthy replied:
There are new test logic added in 2.2 . Can you please the test case issue in this branch?

kalyc replied:
Could remove this skip test attribute once we have default data format set to 'channels_first'  & throwing an error from conv1d with 'channels_last'

kalyc commented on keras/layers/convolutional.py:
```diff
@@ -321,13 +330,18 @@ def __init__(self, filters,
                  kernel_constraint=None,
                  bias_constraint=None,
                  **kwargs):
+        if padding == 'causal':
```
was causal mode tested with channels last? I don't recall having facing any issues with data format when I opened PR for this

sandeep-krishnamurthy replied:
This test case is failing.

kalyc replied:
Please see comment above

kalyc replied:
I tested this locally by pulling your changes. It appears that there's an issue with conv1d with all of the paddings when data format is channels_last. Conv1d works properly for channels_first, with any padding. As discussed, we should update the error message to indicate this issue with conv1d

sandeep-krishnamurthy commented:
Thanks @kalyc  your feedback was very helpful. Fixed it.

roywei commented on examples/babi_memnn.py:
```diff
@@ -27,6 +28,11 @@
 import re
 
 
+if K.backend() == 'mxnet':
```
Without adding this check, our backend is already showing a clear error message:
```
NotImplementedError: MXNet Backend: Sparse operations are not supported yet.
```
So I think we can avoid repeated check, and its on our list of not supported example [issue]
(https://github.com/awslabs/keras-apache-mxnet/issues/79) 

I was adding this check for other examples only if error messaging is not clear or does not indicate root cause.

sandeep-krishnamurthy replied:
makes sense removed.

roywei commented on examples/cifar10_cnn_capsule.py:
```diff
@@ -14,14 +14,18 @@
 from keras.layers import *
 from keras.preprocessing.image import ImageDataGenerator
 
+if K.backend() == 'mxnet':
```
thanks for this catch, I have updated the[ issue list](https://github.com/awslabs/keras-apache-mxnet/issues/79)

sandeep-krishnamurthy commented on LICENSE:
```diff
@@ -16,10 +16,6 @@ All contributions by Amazon:
 
-All contributions by Amazon:
-Copyright (c) 2017 Amazon.com, Inc. or its affiliates
-All rights reserved.
```
extra. There is one above this already.

sandeep-krishnamurthy commented on .travis.yml:
```diff
@@ -60,17 +61,17 @@ install:
 
   # install TensorFlow (CPU version).
-  - pip install tensorflow
+  - pip install tensorflow==1.7
```
https://github.com/keras-team/keras/blob/master/.travis.yml#L60

sandeep-krishnamurthy commented:
* Fixed class_weight ignore issue.
* Fixed accuracy issue with custom loss.
* Added warning message for a workaround if users are using custom loss and slice operation. Linked to issue here - https://github.com/awslabs/keras-apache-mxnet/issues/120
* Updated Conv1D test

I would like to merge this PR to master prepare v 2.2 beta2 build that should be tested with latest MXNet 1.2.1 and make a release. We can then updated dev from master post release. Suggestions?

@roywei @kalyc
